# Germany E-Bay Cars Analysis

Understanding how to navigate the used car market is an invaluable skill for people in the market for a car. I myself will be looking to purchase a car soon (most likely used) so this project piqued my interest. We will explore a dataset of used cars from the website eBay Kleinanzeige. 


* The data is no longer able to be accessed on kaggle, but can found at the following link https://data.world/data-society/used-cars-data.
* The dataset contains 50,000 data points. 
* Data Dictionary:

    - `dateCrawled`:- When this ad was first crawled. All field-values are taken from this date.
    - `name`:- Name of the car.
    - `seller`:- Whether the seller is private or a dealer.
    - `offerType`:- The type of listing
    - `price`:- The price on the ad to sell the car.
    - `abtest`:- Whether the listing is included in an A/B test.
    - `vehicleType`:- The vehicle Type.
    - `yearOfRegistration` - The year in which the car was first registered.
    - `gearbox`:- The transmission type.
    - `powerPS`:- The power of the car in PS.
    - `model`:- The car model name.
    - `kilometer`:- How many kilometers the car has driven.
    - `monthOfRegistration`:- The month in which the car was first registered.
    - `fuelType`:- What type of fuel the car uses.
    - `brand`:- The brand of the car.
    - `notRepairedDamage`:- If the car has a damage which is not yet repaired.
    - `dateCreated`:- The date on which the eBay listing was created.
    - `nrOfPictures`:- The number of pictures in the ad.
    - `postalCode`:- The postal code for the location of the vehicle.
    - `lastSeenOnline`:- When the crawler saw this ad last online.

#### The goal of this project is to clean data and analyze the listings of used cars. Some additional insights that could result from this analysis include more information on used car popularity and which cars typically offer the best value when measured against price for those bargain shoppers. 

### Importing Libraries
I'll need to import **NumPy** to perform quick operations on data. I'll need **Pandas** to work with dataframes so that I can use different datatypes all while using the same concepts that make **NumPy** so valuable mainly **vectorized operations**. 

In [2]:
## Packages

import numpy as np
import pandas as pd


In [3]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [5]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### Observations
Looking at the autos dataset, I can see that there are several issues with the present structure of the data. First of all, there are some instances where numeric data is presented as an object. 

* One such example is the price feature that has the price of the vehicle. There's a special character "$" in that row which will need to be included in the title, but not in the actual values. 

* Another example is the odometer column which has characters for each value which make it an object. The unit of measurement for the odometer should be kept included in the feature title. 

* The date created is also listed as an object instead of a numeric or datetime type. The same is the case for the last seen column. 

### Null Values
Another issue with the dataset is the null values for a variety of features. For modeling and machine learning purposes we need each feature to be equally complete. The following features that have issues are:

* vehicleType
* gearbox
* model
* fuelType
* notRepairedDamage

In [6]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

### Additonal observations
* The dataset has 20 columns(features) and the majority are of the object type which is a string type. 
* There are null values in several categories. However, none of the columns have more than 20% null values. 
* The column names use camel case instead of snakecase which is preferred in Python environments. 

Below, we are to going to fix the column name issues by converting the camelcase of the columns into snakecase while also rewording some of the different column names. 

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

### Reasoning
Getting the data into a format that is optimal for Python will be helpful in our analysis. This includes column lables. Which is why we made the switch from camelcase to snakecase. 

### Basic Data Exploration

Next, we are going to further explore the data to determine what are some other cleaning tasks that need to be done. I will try to find:

* Text columns were all or almost all of the values are the same. These columns can often be dropped because they aren't that useful or have useful information for analysis. 
* Data that is not stored in the proper format to conduct analysis on it. For example, numeric values that aren't stored as numerics, but rather as strings. 

Next,

* I am going to use the `DataFrame.describe(include='all')` method to get information for categorical and numeric columns.
* After that, I will use the  `Series.value_counts()` and `Series.head()` methods to determine which columns might need to be looked at further. 

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

### Takeaways

* The *seller, offer_type, ab_test, gearbox, unrepaired_damage* columns only have a couple of unique values which might not contribute much to the analysis. 
* The *num_photos* column looks a little funny because it has a mean of zero despite having so many values. 

In [10]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

The column has a zero for every column, so this column will be dropped along with the *seller and offer_type columns*. 



In [11]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis = 1)

I didn't drop the *ab_test, gearbox, and unrepaired damage* column, but those contain values or information that might need to be considered by the person looking for a car. Most people want to know if their car was included in an AB test, they also want to know the type of transmission indicated by the *gearbox column*. Finally, I'd say most people want to know what kind of damaged occured to the vehicle and whether or not it has been repaired. 

### Converting numeric data from text

Earlier, I found that the `price` and `odometer` columns were saved as object types which isn't representative of what they are **numerics**. For these two columns I'll remove the non_numeric characters, convert the values to a numeric, and then rename the column names so that type of descriptive information is not lost. 

* The `price` column contains non_numerics like `,` and `$`. The odometer column has `,` and `km` in its values. 

In [12]:
autos["price"] = (autos["price"]
                 .str.replace("$", "")
                 .str.replace(",", "")
                 .astype(int)
                 )
autos.rename({'price': 'price_usd'}, inplace = True, axis = 1)
autos["price_usd"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price_usd, dtype: int64

In [13]:
autos["odometer"] = (autos["odometer"]
                    .str.replace("km", "")
                    .str.replace(",", "")
                    .astype(int)
                    )
autos.rename({'odometer': 'odometer_km'}, inplace = True, axis = 1)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

### Looking for numeric irregularities


In [14]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

### Observations
From what we can tell the distances are rounded to the nearest thousandth. I do notice, that there is a column that's only 5000 which is small in comparioson to the others. Additonally, most of the vehicles as expected have higher mileage counts. People don't usually sell a car quickly after purchasing it. 

In [15]:
autos["odometer_km"].unique().shape

(13,)

In [16]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

It looks like there are 13 unique values for the odometer count which indicates an interval that the car fell between. Additionally, we can confirm the idea before that on average most of the cars had higher mileage counts. 

### Exploring Price
Next, I'll have a look at the price feature to see if there are any irregularities. 

In [17]:
autos["price_usd"].unique().shape

(2357,)

In [18]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [19]:
autos["price_usd"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price_usd, Length: 2357, dtype: int64

In [20]:
autos["price_usd"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

In [21]:
autos["price_usd"].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

From the information above we can see that there quite a few values with very low values for the listing. There are nearly 1,500 listings that are free. 

Additionally, there are some vehicles that have outrageous prices as well with some cars being valued near $1 million. 

EBay is known for being an auction website so in the case of eBay there is a chance that some of the $1 listings are legitimate because the auction might start with $1 as an initial value. Additionally, we are going to remove extremely high values as they don't seem to realistically fall within the values expected of a used car. We'll stop for cars valued over $350,000. 

In [22]:
autos[autos["price_usd"].between(1, 351000)]
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

### Exploring the Date columns

From earlier, we saw that there were five columns that had date information:

- `date_crawled`
- `registration_month`
- `registration_year`
- `ad_created`
- `last_seen`

* The `date_crawled`, `last_seen`, and `ad_created` columns were identified as objects(string) values in pandas. In order to evaluate these columns we will have to convert them to a datetime object. 

* `registration_month` and  `registration_year` are in numeric form. We can use use the `Series.describe()` method to have a look at the distrubution without having to process additional data. 

* **First**, let's take a look and explore each column so we can learn more about the listings. 

In [23]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [24]:
autos["date_crawled"].str[:10]

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 50000, dtype: object

In [25]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [26]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize= True, dropna=False)
 .sort_values()
)

2016-04-07    0.00142
2016-04-06    0.00318
2016-03-18    0.01306
2016-04-05    0.01310
2016-03-06    0.01394
2016-03-13    0.01556
2016-03-05    0.02538
2016-03-24    0.02910
2016-03-16    0.02950
2016-03-27    0.03104
2016-03-17    0.03152
2016-03-25    0.03174
2016-03-31    0.03192
2016-03-10    0.03212
2016-03-23    0.03238
2016-03-26    0.03248
2016-03-11    0.03248
2016-03-22    0.03294
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-30    0.03362
2016-04-01    0.03380
2016-03-15    0.03398
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-19    0.03490
2016-04-02    0.03540
2016-03-07    0.03596
2016-04-04    0.03652
2016-03-14    0.03662
2016-03-12    0.03678
2016-03-21    0.03752
2016-03-20    0.03782
2016-04-03    0.03868
Name: date_crawled, dtype: float64

The site was crawed over a little more than a month and the distribution of the listings seems to be pretty similar over the duration of the crawl. 

In [27]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize = True, dropna= False)
        .sort_index()
        )

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

The crawler recorded the date that it last saw any listing which means that a list may have been removed. This probably means that the car was sold. 

The last few days of the 'last seen' values are quite a bit larger in reference to the rest of the data. Sales seems to be relatively stable. This might indicate that the crawl period was coming to an end. 

In [28]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad creation dates. These usually fall within two months of the listing date, but some are relatively old reaching up to 9 months. 

In [29]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Observations

* The minimum value for the `registration_year` feature is 1000 which doesn't really compute. Cars weren't invented in the year 1000. 

* The maximum value is 9999, thousands of years into the future. There will need to be some corrections made to the `registration_year feature`. 

**First**, we'll have to capture the correct range of dates for the dataset. Considerig the dataset is from 2016, we can't include any vehicles with a registration date above 2016. Choosing the earliest year for to segment the dataset by is also difficult. Cars started to be built in the early 1900's so we'll start with a date of 1900. 

In [39]:
invalid_rows = (~(autos["registration_year"].between(1900, 2016))).sum()
total_rows = autos["registration_year"].shape[0]
percentage_invalid = invalid_rows/total_rows

print('Invalid years make up {:.1f}% of the total data'.format(percentage_invalid *100))

Invalid years make up 3.9% of the total data


The registration years only make up 3.9% of the data, so we can eliminate these records from further analysis. 

In [47]:
# Series.between method to select data that is in between 1900-2016. 


auto = autos[autos["registration_year"].between(1900, 2016)]
years_registered = autos["registration_year"].value_counts(normalize = True).sort_values(ascending= False)

years_registered.head(10)

2000    0.06708
2005    0.06030
1999    0.06000
2004    0.05474
2003    0.05454
2006    0.05416
2001    0.05406
2002    0.05066
1998    0.04906
2007    0.04608
Name: registration_year, dtype: float64

Looking at the data above most of the years are from the past 20 years

### Exploring Price by Brand

In [48]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

Not surprisingly, German manufactured vehicles make up 4 of the top 5 brands. These make up nearly 50% of the total listings. 


There are several vehicle manufacturers that don't make up a significant percentage of the listings available. So what we'll do is limit the listing to brands that represent at least 5% of total listings. 


In [50]:
brand_counts = autos["brand"].value_counts(normalize = True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [53]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price_usd"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 6384,
 'opel': 5106,
 'bmw': 8252,
 'mercedes_benz': 29511,
 'audi': 8965,
 'ford': 7105}

Looking at the top brands above:

* There is a clear price grap. Mercedes, BMW, Audi are the more expensive brands. Mercedes is clearly the most expensive brand. 
* Ford and Opel brands are lower priced than the three just mentioned. 
* Volkwagen splits the Opel and Ford which might indicate the sweet spot for pricing when it comes to consumers. Enough value, but at a fair price. 

### Exploring Mileage



In [54]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

volkswagen        6384
opel              5106
bmw               8252
mercedes_benz    29511
audi              8965
ford              7105
dtype: int64


In [57]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending = False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending = False)

In [60]:
brand_info = pd.DataFrame(mean_mileage, columns = ['mean_milage'])
brand_info

mean_milage
bmw                 132521
mercedes_benz       130886
audi                129643
opel                129298
volkswagen          128955
ford                124131

In [61]:
brand_info["mean_price"] = mean_prices
brand_info.head()

mean_milage  mean_price
bmw                 132521        8252
mercedes_benz       130886       29511
audi                129643        8965
opel                129298        5106
volkswagen          128955        6384

From what we can see by constructing a dataframe with the `mean_mileage` and `mean_price` there isn't as much variety between mileage as there is between prices. The more expensive vehicles tend to have higher mileage, which is probably from the power of the brand. 